Topics/Tech_Stacks/Writing_Clean_React_Code.md

```diff
@@ -0,0 +1,153 @@
+
+  ```jsx
+  src
+  ├── components
+  │   ├── Home
+  │   │   ├── Navbar
+  │   │   │   └── ...
+  │   │   ├── Home.jsx
+  │   │   ├── Home.test.jsx
+  │   │   └── Home.css
+  │   └── Profile
+  │       ├── ProfileElements
+  │       │   └── ...
+  │       ├── Profile.jsx
+  │       ├── Profile.test.jsx
+  │       └── Profile.css
+  ├── App.jsx
+  ├── App.test.jsx
+  └── index.js
+  ```
+
+- **Use container components**: Containers are useful for styling, since they can be used to wrap multiple components and apply flex or grid properties to them. They can also be responsible for fetching data and passing it down to child components. This helps separate the logic of fetching data from the logic of rendering components. For example, a component that renders a user profile page should be broken down into a container component that fetches the user's data and a child component that renders the user's data. This is also known as the [smart and dumb component pattern](https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0), which dates back to class components.
+
+- **Try to use fewer States**: State management is a powerful tool in React, but it can also be overused. It is important to consider whether a component really needs to be stateful. If data to persist across re-renders, but don't want every update to **cause** a re-render, you might be interested in using useRef hooks instead. Moreover, it is easier to reuse stateless components.
+
+One way to reduce the number of states is to group related states into a single state object. For example, a component that fetches user data from a database can store the user's name, email, and bio in a single state object instead of storing each piece of data in a separate state.
+
+```jsx
+// This is a bad example
+const [name, setName] = useState("");
+const [email, setEmail] = useState("");
+const [bio, setBio] = useState("");
+```
+
+```jsx
+// This is a better example
+const [user, setUser] = useState({
+  name: "",
+  email: "",
+  bio: "",
+});
+
+// It is also good to destruct the state object when using it
+const { name, email, bio } = user;
+```
+
+It is important to note that this is not always the best approach, since it can lead to a bloated state object, and its also important for everyone working in the codebase to be aware of the individual states that are grouped together in the state object.
```

Comment:
nit: Typo -"its" should be "it's"

Reply:
You are correct. Thank you for finding these nitpicks!